cmd/utils.go

```diff
@@ -128,42 +145,72 @@ func findCachedCastDNS(deviceName, deviceUuid string) castdns.CastDNSEntry {
 	return CachedDNSEntry{}
 }
 
-func findCastDNS(device, deviceName, deviceUuid string) (castdns.CastDNSEntry, error) {
-	dnsEntries := castdns.FindCastDNSEntries()
-	switch l := len(dnsEntries); l {
-	case 0:
-		return castdns.CastEntry{}, errors.New("no cast dns entries found")
-	default:
-		for _, d := range dnsEntries {
-			if (deviceUuid != "" && d.UUID == deviceUuid) || (deviceName != "" && d.DeviceName == deviceName) || (device != "" && d.Device == device) {
```

Comment:
I think this should stay here instead of the `DeviceMatcher` approach. I like the `DeviceMatcher` approach usually, but in this case the previous approach a bit easier to reason about and simplifies the code quite a bit since there won't ever be more than 3 things to match on.

Reply:
I refactored it, to make more similar to the previous approach. I agree that this is now much clearer:
```go
func deviceMatcher(deviceType, deviceName, deviceUuid string) discovery.DeviceMatcher {
	return func(d *discovery.Device) bool {
		if d == nil {
			return false
		}
		if deviceType != "" && deviceType != d.Type() {
			return false
		}
		if deviceName != "" && deviceName != d.Name() {
			return false
		}
		if deviceUuid != "" && deviceUuid != d.ID() {
			return false
		}
		return true
	}
}
```

What do you think?

Reply:
It is cleaner than before but it is still a bit verbose for what it is doing. I think leave it as it was and remove the `DeviceMatcher` stuff since it is over-complicated for the task it is doing.

Reply:
>  I think leave it as it was and remove the DeviceMatcher stuff

So it means moving the logic from the `discovery.First` function in this file?
The `discovery` package currently does not expose any asynchronous function (the channel logic is hidden, to simplify its API). To implement the `First` function here would require the `discovery` package to expose the underlying channel.

I am not sure what should then be exposed or not...

Maybe can you propose some code, how you would solve this?

Reply:
Yup, I think passing in a channel from here to the `discovery` package (similar to how mdsn and zeroconf take a channel) and then adding the logic for the first functionality here is a better solution. The `discovery` package can just be a thin wrapper around `zeroconf` (or `mdns`) which returns a device in a standardize format and then whoever passes in the channel is responsible for any logic required. What do you think?

Reply:
> What do you think?

I guess it more or less means removing the `discovery` package and implementing its logic directly in `utils.go`.
If you insist on implementing it this way, I would prefer if you could do it yourself (otherwise I am afraid that I won't find enough time for this).

Reply:
I think the `discovery` package is still needed, but it is a provides a common interface for `mdns` or `zeroconf` or anything else. But yes, I think the logic should be in this file since it is very unique to only this problem. That's fine, if you don't think you can get to it feel free to close this PR and I will take a look at implementing it when I have a chance.